File: packages/app-floorplan/src/add-opening-handler.ts
import { LengthMm } from "@toporoom/domain-floorplan";
import type { OpeningKind } from "@toporoom/domain-floorplan";
import type { DocumentStorePort } from "@toporoom/ports";
import { commit, loadDocument, type CommandResult } from "./document-io.js";
import type { GeometryRebuildPolicy } from "./rebuild-policy.js";

export interface AddOpeningCommand {
  documentId: string;
  storeyId: string;
  wallId: string;
  openingId?: string;
  kind: OpeningKind;
  widthMm: number;
  heightMm: number;
  offsetMm: number;
  sillHeightMm: number;
}

export class AddOpeningHandler {
  constructor(
    private readonly store: DocumentStorePort,
    private readonly rebuild: GeometryRebuildPolicy,
  ) {}

  async execute(command: AddOpeningCommand): Promise<CommandResult> {
    const document = await loadDocument(this.store, command.documentId);
    document.addOpening({
      storeyId: command.storeyId,
      wallId: command.wallId,
      id: command.openingId,
      kind: command.kind,
      width: LengthMm.of(command.widthMm),
      height: LengthMm.of(command.heightMm),
      offsetAlongWall: LengthMm.of(command.offsetMm),
      sillHeight: LengthMm.of(command.sillHeightMm),
    });
    return commit(this.store, this.rebuild, document);
  }
}

File: core/src/app/add_opening_handler.cpp
#include "toporoom/app/add_opening_handler.hpp"

#include "toporoom/domain/floor_plan_document.hpp"
#include "toporoom/domain/length_mm.hpp"

namespace toporoom::app {

CommandResult AddOpeningHandler::execute(const AddOpeningCommand& command) {
  auto document = load_document(store_, command.document_id);
  domain::AddOpeningProps props;
  props.storey_id = command.storey_id;
  props.wall_id = command.wall_id;
  props.id = command.opening_id;
  props.kind = command.kind;
  props.width = domain::LengthMm::of(command.width_mm);
  props.height = domain::LengthMm::of(command.height_mm);
  props.offset_along_wall = domain::LengthMm::of(command.offset_mm);
  props.sill_height = domain::LengthMm::of(command.sill_height_mm);
  document.add_opening(std::move(props));
  return commit(store_, rebuild_, document);
}

}  // namespace toporoom::app
